src/hash.rs

```diff
@@ -88,6 +89,58 @@ pub struct PoseidonCache<F: LurkField> {
     pub constants: HashConstants<F>,
 }
 
+#[derive(Default, Debug)]
+pub struct InversePoseidonCache<F: LurkField> {
+    a3: HashMap<FWrap<F>, [F; 3]>,
+    a4: HashMap<FWrap<F>, [F; 4]>,
+    a6: HashMap<FWrap<F>, [F; 6]>,
+    a8: HashMap<FWrap<F>, [F; 8]>,
+
+    pub constants: HashConstants<F>,
+}
+
+impl<F: LurkField> InversePoseidonCache<F> {
+    pub fn get<const ARITY: usize>(&self, key: &FWrap<F>) -> Option<&[F; ARITY]> {
+        macro_rules! get {
+            ($name:ident, $n: expr) => {{
+                let preimage = self.$name.get(key);
+                if let Some(p) = preimage {
+                    assert_eq!(ARITY, $n);
+                    Some(unsafe { std::mem::transmute::<&[F; $n], &[F; ARITY]>(p) })
+                } else {
+                    None
+                }
+            }};
+        }
+
+        match ARITY {
+            3 => get!(a3, 3),
+            4 => get!(a4, 4),
+            6 => get!(a6, 6),
+            8 => get!(a8, 8),
+            _ => unreachable!(),
```

Comment:
`unimplemented!()`?

Reply:
The idea is that this code is actually statically unreachable in any valid call.